internal/confmapprovider/discovery/discoverer.go

```diff
@@ -0,0 +1,525 @@
+
+	"github.com/signalfx/splunk-otel-collector/internal/common/discovery"
+	"github.com/signalfx/splunk-otel-collector/internal/components"
+	"github.com/signalfx/splunk-otel-collector/internal/receiver/discoveryreceiver"
+	"github.com/signalfx/splunk-otel-collector/internal/version"
+)
+
+const durationEnvVar = "SPLUNK_DISCOVERY_DURATION"
+
+// discoverer provides the mechanism for a "preflight" collector service
+// that will stand up the observers and discovery receivers based on the .discovery.yaml
+// contents of the config dir, acting as a log consumer to determine which
+// of the underlying receivers were successfully discovered by the
+// discovery receiver from its emitted log records.
+type discoverer struct {
+	factories           component.Factories
+	expandConverter     confmap.Converter
+	configs             map[string]*Config
+	extensions          map[config.ComponentID]component.Extension
+	logger              *zap.Logger
+	discoveredReceivers map[config.ComponentID]discovery.StatusType
+	discoveredConfig    map[config.ComponentID]map[string]any
+	discoveredObservers map[config.ComponentID]discovery.StatusType
+	info                component.BuildInfo
+	duration            time.Duration
+	mu                  sync.Mutex
+}
+
+func newDiscoverer(logger *zap.Logger) (*discoverer, error) {
+	info := component.BuildInfo{
+		Command: "discovery",
+		Version: version.Version,
+	}
+	duration := 10 * time.Second
+	if d, ok := os.LookupEnv(durationEnvVar); ok {
+		if dur, err := time.ParseDuration(d); err != nil {
+			logger.Warn("Invalid SPLUNK_DISCOVERY_DURATION. Using default of 10s", zap.String("duration", d))
+		} else {
+			duration = dur
+		}
+	}
+
+	factories, err := components.Get()
+	if err != nil {
+		return (*discoverer)(nil), err
+	}
+
+	m := &discoverer{
+		logger:              logger,
+		info:                info,
+		factories:           factories,
+		extensions:          map[config.ComponentID]component.Extension{},
+		configs:             map[string]*Config{},
+		duration:            duration,
+		mu:                  sync.Mutex{},
+		expandConverter:     expandconverter.New(),
+		discoveredReceivers: map[config.ComponentID]discovery.StatusType{},
+		discoveredConfig:    map[config.ComponentID]map[string]any{},
+		discoveredObservers: map[config.ComponentID]discovery.StatusType{},
+	}
+	return m, nil
+}
+
+// discover will create all .discovery.yaml components, start them, wait the configured
+// duration, and tear them down before returning the discovery config.
+func (d *discoverer) discover(cfg *Config) (map[string]any, error) {
+	discoveryReceivers, discoveryObservers, err := d.createDiscoveryReceiversAndObservers(cfg)
+	if err != nil {
+		d.logger.Error("failed preparing discovery components", zap.Error(err))
+		return nil, err
+	}
+
+	for observerID, observer := range discoveryObservers {
+		d.logger.Debug(fmt.Sprintf("starting %s", observerID.String()))
+		if e := observer.Start(context.Background(), d); e != nil {
```

Comment:
Could we take a context with a timeout there?

Reply:
good call, updated for all other the component methods